system/mp/test/mptest.cpp

```diff
@@ -521,127 +523,214 @@ void MultiTest(ICommunicator *_comm)
+    PrintLog("MPTEST: MPRing %d send to rank %d", myrank, next);
+    mpicomm->send(smb,next,MPTAG_TEST);
+
+    while (1)
     {
-        if (myrank == 0)
-        {
-            left = 1;
-            right = 1;
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
-
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
-        }
-        else
+        rank_t r;
+        rmb.clear();
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d recv from rank %d", myrank, prev);
+        mpicomm->recv(rmb,prev,MPTAG_TEST,&r);
+        rmb.read(iters);
+
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d iters = %d", myrank, iters);
+
+        if (iters-- == 0)
+            break;
+
+        // shift source and dest around the ring ...
+
+        for (int i=0; i<numranks; i++)
         {
-            left = 0;
-            right = 0;
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
-
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
+            int n1 = na[i];
+            n1++;
+            if (n1 == numranks)
+                n1 = 0;
+            if (n1 == i)
+                n1++;
+            if (n1 == numranks)
+                n1 = 0;
+            na[i] = n1;
+            pa[na[i]] = i;
         }
+
+        next = na[myrank];
+        prev = pa[myrank];
+
+        smb.clear();
+        smb.append(iters);
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d send to rank %d", myrank, next);
+        mpicomm->send(smb,next,MPTAG_TEST);
     }
-    else if (numranks > 2)
+
+    PrintLog("MPTEST: MPRing complete");
+
+    mpicomm->barrier();
+
+    delete [] na, pa;
+
+    return;
+}
+
+#define MSGLEN 1048576
+
+void MPAlltoAll(IGroup *group, ICommunicator *_mpicomm, int buffsize=0, int iters=0)
+{
+    int myrank = group->rank();
+    int numranks = group->ordinality();
+
+    if (buffsize == 0)
+        buffsize = MSGLEN;
+    if (iters == 0)
+        iters = 1000;
+    if (iters < 10)
+        iters = 10;
+
+    PrintLog("MPTEST: MPAlltoAll myrank=%d numranks=%d buffsize=%d iters=%d", myrank, numranks, buffsize, iters);
+
+    // ---------
+
+    class Sender : public Thread
     {
-        int m = 0;
-        while (m < (numranks - 1))
+    public:
+        Owned<ICommunicator> mpicomm;
+        int numranks;
+        int myrank;
+        int buffsize;
+        int iters;
+        char *buff;
+        Sender(ICommunicator *_mpicomm, int _numranks, int _myrank, int _buffsize, int _iters)
```

Comment:
minor: better types variables/members.. unsigned for nunranks, rank_t for myrank, size32_t for buffsize, unsigned for iters